### PR TITLE
fix: filetype, filesize in library, magic shelf, opds

### PR DESCRIPTION
**Pull request by WorldTeacher**
## 📝 Description

This fixes the filetype and filesize filters in 
- OPDS
- Library View
- Magic Shelves

**Linked Issue:** Fixes #2651

> **Required.** Every PR must reference an approved issue. If no issue exists, [open one](https://github.com/booklore-app/booklore/issues/new) and wait for maintainer approval before submitting a PR. Unsolicited PRs without a linked issue will be closed.

## 🏷️ Type of Change

- [x] Bug fix
- [ ] New feature
- [ ] Enhancement to existing feature
- [ ] Refactor (no behavior change)
- [ ] Breaking change (existing functionality affected)
- [ ] Documentation update

## 🔧 Changes

- changed filesize:  use `book.primaryFile?.fileSizeKb` instead of `book.fileSizeKb`
- changed filetype: use `book.primaryFile?.bookType` instead of `book['bookType']`
- added filesize for OPDS feed as `feed.append(" length=\"").append(bookFile.getFileSizeKb() * 1024).append("\"");`
- changed the BookRuleEvaluatorService FILE_SIZE to fix filesize 
- added file size ranges: 100-250 MB, 2-5GB, which were missing from the ranges
- added / updated tests to test for filesize and filetype 
## 🧪 Testing (MANDATORY)

> **PRs without this section filled out will be closed.** "Tests pass" or "Tested locally" is not sufficient. You must provide specifics.

**Manual testing steps you performed:**

1. ran local tests for front- and backend
2. verified a docker image can be run
3. tested all filters in the library, magic shelf and opds feed

**Regression testing:**

- as no other areas were touched, existing features did not change

**Edge cases covered:**

- none / unclear

**Test output:**

<details>
<summary>Backend test output (<code>./gradlew test</code>)</summary>

```
Starting a Gradle Daemon (subsequent builds will be faster)

> Task :compileJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
Management of bidirectional association persistent attributes is deprecated and will be removed. Set the value to 'false' to get rid of this warning

> Task :compileTestJava
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
OpenJDK 64-Bit Server VM warning: Sharing is only supported for boot loader classes because bootstrap classpath has been appended
2026-02-19T20:45:49.271+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : Stopping...
2026-02-19T20:45:49.271+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : BrokerAvailabilityEvent[available=false, SimpleBrokerMessageHandler [org.springframework.messaging.simp.broker.DefaultSubscriptionRegistry@22c2919b]]
2026-02-19T20:45:49.271+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.s.m.s.b.SimpleBrokerMessageHandler     : Stopped.
2026-02-19T20:45:49.279+01:00  INFO 2179454 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-02-19T20:45:49.298+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-02-19T20:45:49.598+01:00  INFO 2179454 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-02-19T20:45:49.599+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-02-19T20:45:49.599+01:00  WARN 2179454 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-02-19T20:45:49.599+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-02-19T20:45:49.603+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-02-19T20:45:49.613+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-02-19T20:45:49.613+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-02-19T20:45:49.617+01:00  INFO 2179454 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-02-19T20:45:49.617+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-02-19T20:45:49.617+01:00  INFO 2179454 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-02-19T20:45:49.618+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-02-19T20:45:49.618+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-02-19T20:45:49.618+01:00  WARN 2179454 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-02-19T20:45:49.623+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-02-19T20:45:49.625+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-02-19T20:45:49.625+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-02-19T20:45:49.628+01:00  INFO 2179454 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-02-19T20:45:49.628+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-02-19T20:45:49.628+01:00  INFO 2179454 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-02-19T20:45:49.629+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-02-19T20:45:49.629+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-02-19T20:45:49.629+01:00  WARN 2179454 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-02-19T20:45:49.634+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-02-19T20:45:49.638+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-02-19T20:45:49.638+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-02-19T20:45:49.642+01:00  INFO 2179454 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-02-19T20:45:49.642+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-02-19T20:45:49.642+01:00  INFO 2179454 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-02-19T20:45:49.643+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-02-19T20:45:49.643+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-02-19T20:45:49.643+01:00  WARN 2179454 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-02-19T20:45:49.648+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-02-19T20:45:49.652+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-02-19T20:45:49.653+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-02-19T20:45:49.656+01:00  INFO 2179454 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-02-19T20:45:49.656+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-02-19T20:45:49.656+01:00  INFO 2179454 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-02-19T20:45:49.657+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-02-19T20:45:49.657+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-02-19T20:45:49.657+01:00  WARN 2179454 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-02-19T20:45:49.662+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-02-19T20:45:49.666+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-02-19T20:45:49.666+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.
2026-02-19T20:45:49.670+01:00  INFO 2179454 --- [booklore-api] [opFolderWatcher] o.b.s.b.BookdropMonitoringService        : Bookdrop monitor thread interrupted
2026-02-19T20:45:49.685+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.b.BookdropMonitoringService        : Stopped bookdrop folder monitor
2026-02-19T20:45:49.685+01:00  INFO 2179454 --- [booklore-api] [opFileProcessor] o.b.s.b.BookdropEventHandlerService      : File processing thread interrupted, shutting down.
2026-02-19T20:45:49.686+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.monitoring.MonitoringService       : Shutting down monitoring service...
2026-02-19T20:45:49.686+01:00  WARN 2179454 --- [booklore-api] [        async-0] o.b.service.monitoring.MonitoringTask    : WatchService has been closed. Stopping monitoring.
2026-02-19T20:45:49.686+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] o.b.s.watcher.LibraryFileEventProcessor  : Shutting down LibraryFileEventProcessor...
2026-02-19T20:45:49.690+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] j.LocalContainerEntityManagerFactoryBean : Closing JPA EntityManagerFactory for persistence unit 'default'
2026-02-19T20:45:49.693+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown initiated...
2026-02-19T20:45:49.694+01:00  INFO 2179454 --- [booklore-api] [ionShutdownHook] com.zaxxer.hikari.HikariDataSource       : BookloreHikariPool - Shutdown completed.

[Incubating] Problems report is available at: file:///home/user/GitHub/booklore/booklore-api/build/reports/problems/problems-report.html

BUILD SUCCESSFUL in 1m 4s
6 actionable tasks: 5 executed, 1 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.3.1/userguide/configuration_cache_enabling.html
```

</details>

<details>
<summary>Frontend test output (<code>ng test</code>)</summary>

```Using Vitest configuration file: /home/user/GitHub/booklore/booklore-ui/vitest-base.config.ts
Initial chunk files                                                            | Names                                                                       |  Raw size
spec-app-features-magic-shelf-component-magic-shelf-component.js               | spec-app-features-magic-shelf-component-magic-shelf-component               |   1.09 MB | 
spec-app-features-magic-shelf-service-book-rule-evaluator.service.js           | spec-app-features-magic-shelf-service-book-rule-evaluator.service           |  64.04 kB | 
chunk-7GUG5YIO.js                                                              | -                                                                           |  60.70 kB | 
spec-app-app.component.js                                                      | spec-app-app.component                                                      |  58.63 kB | 
styles.css                                                                     | styles                                                                      |  30.99 kB | 
chunk-GFQBFBNK.js                                                              | -                                                                           |  24.24 kB | 
spec-app-features-magic-shelf-service-book-rule-evaluator-metadata-presence.js | spec-app-features-magic-shelf-service-book-rule-evaluator-metadata-presence |  23.21 kB | 
chunk-EREA7DEY.js                                                              | -                                                                           |  15.59 kB | 
spec-app-features-magic-shelf-service-magic-shelf-utils.js                     | spec-app-features-magic-shelf-service-magic-shelf-utils                     |  11.36 kB | 
spec-app-core-security-auth-initializer.js                                     | spec-app-core-security-auth-initializer                                     |   8.69 kB | 
chunk-PUIDBMXW.js                                                              | -                                                                           |   6.10 kB | 
chunk-TRPOYXPA.js                                                              | -                                                                           |   4.86 kB | 
chunk-3J3QIHZF.js                                                              | -                                                                           |   2.19 kB | 
chunk-HPNCK62B.js                                                              | -                                                                           |   1.33 kB | 
init-testbed.js                                                                | init-testbed                                                                |   1.27 kB | 
spec-app-app.js                                                                | spec-app-app                                                                | 202 bytes | 
polyfills.js                                                                   | polyfills                                                                   | 121 bytes | 

                                                                               | Initial total                                                               |   1.41 MB

Application bundle generation complete. [3.842 seconds] - 2026-02-19T19:43:59.482Z

Watch mode enabled. Watching for file changes...

 DEV  v4.0.18 /home/user/GitHub/booklore/booklore-ui

 ✓  booklore  src/app/app.spec.ts (1 test) 2ms
 ✓  booklore  src/app/features/magic-shelf/service/magic-shelf-utils.spec.ts (46 tests) 14ms
 ✓  booklore  src/app/core/security/auth-initializer.spec.ts (2 tests) 16ms
 ✓  booklore  src/app/features/magic-shelf/service/book-rule-evaluator-metadata-presence.spec.ts (85 tests) 88ms
 ✓  booklore  src/app/features/magic-shelf/service/book-rule-evaluator.service.spec.ts (195 tests) 173ms
 ✓  booklore  src/app/app.component.spec.ts (7 tests) 195ms
 ✓  booklore  src/app/features/magic-shelf/component/magic-shelf-component.spec.ts (73 tests) 333ms

 Test Files  7 passed (7)
      Tests  409 passed (409)
   Start at  20:43:59
   Duration  1.95s (transform 875ms, setup 1.15s, import 3.34s, tests 821ms, environment 2.30s)

JUNIT report written to /home/user/GitHub/booklore/booklore-ui/test-results/vitest-results.xml
```

</details>

## 📸 Screen Recording / Screenshots (MANDATORY)

> Every PR must include a **screen recording or screenshots** showing the change working end-to-end in a running local instance (both backend and frontend). This means you must have actually built, run, and tested the code yourself. PRs without visual proof will be closed without review.

previous state can be seen in the linked issue

after applying the PR:
library:
image
magic shelf (both had 0 before:
image
OPDS:
image

---

## ✅ Pre-Submission Checklist

> **All boxes must be checked before requesting review.** Incomplete PRs will be closed without review. No exceptions.

- [x] This PR is linked to an approved issue
- [x] Code follows project style guidelines and conventions
- [x] Branch is up to date with `develop` (merge conflicts resolved)
- [x] I ran the full stack locally (backend + frontend + database) and verified the change works
- [x] Automated tests added or updated to cover changes (backend **and** frontend)
- [x] All tests pass locally and output is pasted above
- [x] Screen recording or screenshots are attached above proving the change works
- [x] PR is a single focused change (one bug fix OR one feature, not multiple unrelated changes)
- [x] PR is reasonably scoped (PRs over 1000+ changed lines will be closed, split into smaller PRs)
- [x] No unsolicited refactors, cleanups, or "improvements" are bundled in
- [ ] Flyway migration versioning is correct _(if schema was modified)_
- [ ] Documentation PR submitted to [booklore-docs](https://github.com/booklore-app/booklore-docs) _(if user-facing changes)_

### 🤖 AI-Assisted Contributions

> **If any part of this PR was generated or assisted by AI tools (Copilot, Claude, ChatGPT, etc.), all items below are mandatory.** You are fully responsible for every line you submit. "The AI wrote it" is not an excuse, and AI-generated PRs that clearly haven't been reviewed are the #1 reason PRs get closed.

- [ ] I have read and understand every line of this PR and can explain any part of it during review
- [ ] I personally ran the code and verified it works (not just trusted the AI's output)
- [ ] PR is scoped to a single logical change, not a dump of everything the AI suggested
- [ ] Tests validate actual behavior, not just coverage (AI-generated tests often assert nothing meaningful)
- [ ] No dead code, placeholder comments, `TODO`s, or unused scaffolding left behind by AI
- [ ] I did not submit refactors, style changes, or "improvements" the AI suggested beyond the scope of the issue

---

### CODE REVIEW

**Comment by WorldTeacher** [comment]
@acx10 I'll take a look at the review tmrw

**Comment by WorldTeacher** [comment]
@acx10 should be ready

File: booklore-api/src/main/java/org/booklore/service/BookRuleEvaluatorService.java [hunk 1]
```diff
@@ -35,19 +35,22 @@ public class BookRuleEvaluatorService {
     public Specification<BookEntity> toSpecification(GroupRule groupRule, Long userId) {
         return (root, query, cb) -> {
             Join<BookEntity, UserBookProgressEntity> progressJoin = root.join("userBookProgress", JoinType.LEFT);
+            Join<BookEntity, BookFileEntity> bookFileJoin = root.join("bookFiles", JoinType.LEFT);
+            bookFileJoin.on(cb.isTrue(bookFileJoin.get("isBookFormat")));
+            query.distinct(true);
```
Comment (acx10) [inline booklore-api/src/main/java/org/booklore/service/BookRuleEvaluatorService.java:43]: This join and `DISTINCT` are added unconditionally for every magic shelf query, even when no rule uses `FILE_SIZE` or `FILE_TYPE`. On large libraries that could add up for rules that only filter on title, author, read status, etc. Worth deferring the join to only when those fields are actually referenced?

File: booklore-api/src/main/java/org/booklore/service/BookRuleEvaluatorService.java [hunk 2]
```diff
@@ -648,48 +653,71 @@ private Predicate buildIsEmpty(Rule rule, CriteriaQuery<?> query, CriteriaBuilde
...
     }
 
-    private Predicate buildIncludesAny(Rule rule, CriteriaBuilder cb, Root<BookEntity> root, Join<BookEntity, UserBookProgressEntity> progressJoin) {
+    private Predicate buildIncludesAny(Rule rule, CriteriaBuilder cb, Root<BookEntity> root, Join<BookEntity, UserBookProgressEntity> progressJoin, Join<BookEntity, BookFileEntity> bookFileJoin) {
         List<String> ruleList = toStringList(rule.getValue());
+        
+        // Map file type values for consistency with database enum values
+        if (rule.getField() == RuleField.FILE_TYPE) {
+            ruleList = ruleList.stream()
+                    .map(this::mapFileTypeValue)
```
Comment (acx10) [inline booklore-api/src/main/java/org/booklore/service/BookRuleEvaluatorService.java]: This file type mapping block is copy-pasted three times (here, line 685, line 702). Could be a small helper to reduce duplication.

File: booklore-ui/src/app/features/book/components/book-browser/book-filter/book-filter.config.ts [hunk 3]
```diff
@@ -79,10 +79,12 @@ export const FILE_SIZE_RANGES: readonly RangeConfig[] = [
...
-  {id: 4, label: '250–500 MB', min: 256000, max: 512000, sortIndex: 4},
-  {id: 5, label: '0.5–1 GB', min: 512000, max: 1048576, sortIndex: 5},
-  {id: 6, label: '1–2 GB', min: 1048576, max: 2097152, sortIndex: 6},
-  {id: 7, label: '5+ GB', min: 5242880, max: Infinity, sortIndex: 7}
+  {id: 4, label: '250–500 MB', min: 256000, max: 512000, sortIndex: 5},
+  {id: 5, label: '0.5–1 GB', min: 512000, max: 1048576, sortIndex: 6},
+  {id: 6, label: '1–2 GB', min: 1048576, max: 2097152, sortIndex: 7},
+  {id: 7, label: '5+ GB', min: 5242880, max: Infinity, sortIndex: 9},
+  {id: 8, label: '100–250 MB', min: 102400, max: 256000, sortIndex: 4},
+  {id: 9, label: '2–5 GB', min: 2097152, max: 5242880, sortIndex: 8}
```
Comment (acx10) [inline booklore-ui/src/app/features/book/components/book-browser/book-filter/book-filter.config.ts]: New ranges are appended at the end instead of inserted in their natural sort position in the array. Not a bug since `sortIndex` handles display ordering, but easier to read if placed in order.

**Comment by WorldTeacher** [comment]
@acx10  this should be ready

**Comment by WorldTeacher** [comment]
Closed due to differences in opinion of open source and how OSS should be handled
